### PR TITLE
Break anonymity

### PR DESCRIPTION
**Pull request by whit537**
This PR breaks anonymity with a double opt-in, closing #236.

 - [x] Design privacy permissions UI.
 - [x] Add db fields.
 - [x] Update json endpoint(s).
 - [x] Wire up js.
 - [ ] Review interaction of `is_singular` with `anonymous_receiving`.
 - [ ] Give givers (I guess I'm starting to call them supporters on this PR?) UI to divulge themselves if the receiver allows it. This should go in the tip CTA in the sidebar.
 - [ ] Update payday (etc.?) to set `transfers.is_anonymous` appropriately (double opt-in).
 - [ ] Fill out the Receiving page with a listing of everyone who has given what over time.
 - [ ] Add column(s) on the Giving page to indicate identity divulging status.
 - [ ] Update history page to include details as appropriate.
 - [ ] Wire up behavior when newly requiring non-anonymity: I guess we should prompt and then clear existing tips from people that don't have `tell_supporting`.

### CODE REVIEW

**Comment by whit537** [comment]
New knob on Account page:

screen shot 2015-01-12 at 11 40 26 pm

Next steps:

 - [ ] Give givers (I guess I'm starting to call them supporters on this PR?) UI to divulge themselves if the receiver allows it. This should go in the tip CTA in the sidebar.
 - [ ] Update payday (etc.?) to set `transfers.is_anonymous` appropriately (double opt-in).
 - [ ] Fill out the Receiving page with a listing of everyone who has given what over time.

**Comment by whit537** [comment]
I suggest we broaden the "Hide total {giving,receiving} from others" to encompass specific {giving,receiving} as well.

**Comment by whit537** [comment]
P.S. I started this PR because both @weallhave24hrs and @renaantosz need it to really start using Gratipay in earnest, and they're the ones whose use cases I'm personally prioritizing right now. In other words I'm personally shifting from the ["Pay Your Heroes" roadmap](https://medium.com/inside-gratipay/roadmap-to-heroes-bcbd51c63b11), which was focused on the "individual giving to another individual" case, to a new roadmap focused on the "individual giving to a business" use case. This is based on the retreat, which I still need to blog about.

**Comment by seanlinsley** [comment]
+1 in general. I'll work on reviewing this tomorrow

!m @whit537

File: www/%username/account/index.html.spt [hunk 1]
```diff
@@ -202,6 +202,15 @@ emails = participant.get_emails()
             {% endif %}
         </p>
 
+        <h2>{{ _("Supporter Identity") }}</h2>
+        <p>
+            <label class="anonymous-supporters">
+                <input type="checkbox"
+                    {% if not participant.anonymous_supporters %}checked="true"{% endif %} />
+                    {{ _("Allow my supporters to divulge themselves.") }}
```
Comment (clone1018) [inline www/%username/account/index.html.spt]: I don't feel like this is very descriptive. If you want to keep it we should at least have hover state help text.

**Comment by benhc123** [comment]
Completely agree with @clone1018, this functionality needs to be explained better in the labels.
Overall though, a great function. !m @whit537!

File: www/%username/account/index.html.spt (hunk 1 above)
Comment (rummik) [inline www/%username/account/index.html.spt]: Just a thought:

Allow anyone:
  - [x] To see how much I give
  - [x] To see how much I receive
  - [ ] To search for me

Allow people I support:
  - [ ] To see I support them
  - [ ] To show I support them on their profile
  - [ ] To be shown on my profile

Allow people that support me:
  - [ ] To let me see they support me
  - [ ] To show they support me on their profile
  - [ ] To be shown on my profile

**Comment by whit537** [comment]
I printed out and am rereading #236. The full implementation of #236 gets complicated. On this PR I want to focus on the minimum necessary for @weallhave24hrs @renaantosz, though I want to do it in a way that makes sense in terms of the bigger picture on #236, so that's why I'm rereading that.

**Comment by whit537** [comment]
I also want to look at the specific use cases mentioned on #236. I'm thinking of @jdorfman @jedgar @ielectric etc.

**Comment by iElectric** [comment]
It covers the use case for Kiberpipa.

**Comment by whit537** [comment]
@iElectric Do you need to know how much someone is giving, or just that they're giving? Does it need to be public that they're giving to you or is it enough if you can see that info?

**Comment by chrisdev** [comment]
@whit537 I've got a client https://gratipay.com/birdsongtt/  that is a government registered charity. They would like to give out receipts to donors for tax purposes. Knowing their donors is a prerequisite for this.
I've got a couple more like this coming on stream in the next couple of months. Hoping that this change fits this use case

**Comment by whit537** [comment]
@chrisdev Do they need amounts given, or just identity of givers?

**Comment by iElectric** [comment]
@whit537 
as Kiberpipa
We should see how much is someone donating to us
So that we can account for member fees.

**Comment by whit537** [comment]
@iElectric Okay, thanks for the input.

**Comment by chrisdev** [comment]
@whit537 ideally we need names and amount. But as a second best option the name is good enough

**Comment by whit537** [comment]
@chrisdev Okay, it sounds like @iElectric needs amount, and I know that @renaantosz does as well. So let's consider amount a requirement for this PR to cover your use cases.

**Comment by chrisdev** [comment]
:dancer:  @whit537

**Comment by whit537** [comment]
Gosh, this is a complicated matrix. It's even more complicated than @balupton's helpful matrix at https://github.com/gratipay/gratipay.com/issues/236#issuecomment-24893623, because for @chrisdev's charities case to work we actually need the ability to *require* non-anonymity. I.e., "Only accept tips from people who are willing to divulge their identity to me."

**Comment by whit537** [comment]
There's which side of the transaction you're on: 1) giver, 2) receiver.

There's what piece of info is shared: 1) identity, 2) amount, 3) reason.

There's where it's shared: 1) privately on the other's profile, 2) publicly on other's profile, 3) publicly elsewhere (community page? third-party site?).

There's sharing restrictions: only allow givers with XYZ configuration.

**Comment by whit537** [comment]
We don't have tip reasons/messages yet and we shouldn't add that on this PR, so let's take that out of the consideration for now.

**Comment by whit537** [comment]
Rebased on master.

**Comment by whit537** [comment]
### Privacy

Allow everyone to see:
  - [x] my profile on search results
  - [x] the total amount I receive
  - [x] the total amount I give
  - [ ] who my supporters are
  - [ ] whom I support

Allow people I support to see:
(You can override this for each payment you set up.)
  - [x] who I am
  - [x] the amount I pay them

- [ ] Require my supporters to to show me who they are and how much they pay me.
- [ ] Allow my supporters to to show me who they are and how much they pay me.

**Comment by whit537** [comment]
### Privacy

Allow everyone to see:
  - [x] my profile on search results
  - [x] the total amount I receive
  - [x] the total amount I give
  - [ ] who my supporters are (but not the specific amounts)
  - [ ] whom I support (but not the specific amounts)

Allow people I support to see:
(You can override this for each payment you set up.)
  - [x] who I am
  - [x] the amount I pay them

Who are my supporters?

- [x] I don't want to know who my supporters are.
- [ ] Allow my supporters to show me who they are and how much they pay me.
- [ ] Require my supporters to show me who they are and how much they pay me.

**Comment by whit537** [comment]
Okay! Been going around and around the horn here at Blue Canary with @rummik @kaguillera @dmk246 @renaantosz @eileenford. https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71918901 represents the consensus we've come to over here about the privacy options we want to present to users with this PR.

**Comment by whit537** [comment]
Here's a local iteration that was something we looked at during our discussion:

screen shot 2015-01-28 at 4 44 42 pm

**Comment by clone1018** [comment]
Overload!

**Comment by whit537** [comment]
> Overload!

@clone1018 Yeah, we moved beyond that iteration but I wanted to post it since it was part of our convo.

Here are the four options we looked at:

  1. https://github.com/gratipay/gratipay.com/issues/236#issuecomment-24893774
  1. https://github.com/gratipay/gratipay.com/pull/3112#discussion_r23107496
  1. https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71922937
  1. https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71918901

We analysed them according to the analysis at https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71732435. We landed on option 4.

Some points:

 - We decided to prune the possibility of displaying specific payment amounts publicly, because it didn't seem to be an option that would be popular, and pruning it reduced cognitive/UX overhead.
 - We talked a fair bit about framing these options as opt-in vs. opt-out ("Share"/"Show"/"Allow" vs. "Hide"/"Deny"/"Disallow"). We decided to frame as opt-in because of habituation from other sites (Facebook, Google+), but to use "Allow" instead of "Share" to try to hint more at user control of the experience.
 - The "Who are my supporters?" section would be a radio. The rest would be checkboxes.
 - It would be a bit odd to check "who my supporters are" without also checking "Allow my supporters ...," but at this point I think we should allow that combo (would show identity but not amount on my receiving page, both publicly and to myself).

**Comment by whit537** [comment]
Also:

 - We decided not to tackle the "don't allow anyone to support me who also supports XYZ" case. @rummik's suggestion was that we handle that as part of a "block user" implementation (cf #3060).

**Comment by whit537** [comment]
@iElectric @chrisdev Can I ask you to take a look at https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71918901 and let us know if that set of  permissions will give you what you need for Kiberpipa and Birdsong?

**Comment by whit537** [comment]
One thing that's not possible with https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71918901 is allowing other supporters to see their fellow supporters, without allowing *everyone* to also see who the supporters are. Let's let that use-case go for now since I believe the initial demand for this feature doesn't require it.

**Comment by whit537** [comment]
Hey @wrought @sudoroom, may I loop you in on this conversation? We're working on enabling Gratipay receivers to either allow or require their supporters to divulge who they are and what amount they give. **tl;dr** Is this feature of interest to you, and will the implementation prototyped at https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71918901 give you what you need in this regard?

**Comment by whit537** [comment]
Another case I don't think is handled by https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71918901 is if a company gives to an individual and requires that the individual they're paying know that the payment is coming from their company. So far on #1199 we've been able to avoid this for the case or two where we've hacked an invoice, but as we build out invoices we may run into resistance around the anonymity of our current self-billing plan.

**Comment by timothyfcook** [comment]
awesome!

**Comment by whit537** [comment]
I simplified the "Who?" section by making it entirely about both identity *and* amount, rather than have the second two options include amount but not the first. This rationalizes the situation in point 4 of https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-71926749.

From:

screen shot 2015-01-30 at 10 32 21 am

To:

screen shot 2015-01-30 at 10 32 04 am

**Comment by whit537** [comment]
Current status:

screen shot 2015-01-30 at 10 49 19 am

**Comment by whit537** [comment]
screen shot 2015-01-30 at 11 02 18 am

**Comment by whit537** [comment]
Privacy preferences UI is ready for review. Now to make the preferences actually change things. :-)

**Comment by whit537** [comment]
Do we need "who I support" and "who supports" me to be overridable per-tip? We might.

**Comment by whit537** [comment]
Is this what we need, then?

screen shot 2015-01-30 at 3 19 35 pm

**Comment by whit537** [comment]
The build is breaking at Travis because we don't have the json type in Postgres?

**Comment by whit537** [comment]
Restarted, now seeing failures properly (due to inversion of privacy booleans). Also seeing failures locally.

**Comment by rummik** [comment]
I'm not sure on the `Who pays you what, exactly?` text, since it feels like the wording is too different from the above options.  I'd suggest something more along the lines of:

Information about your supporters:
_Who they are, and how much they pay you._
- [x] Do not show me information about my supporters.
- [ ] Allow my supporters to choose.
- [ ] Require my supporters to tell me.

**Comment by whit537** [comment]
How about this for the tip override UI?

screen shot 2015-02-02 at 11 13 39 am

**Comment by whit537** [comment]
> the wording is too different from the above options

This section *is* different from the above options. If anything, we should set it off with a different heading. We raised but didn't settle the question of whether this is even properly a "privacy" option. It's not about my privacy, it's about my supporter's privacy. "Sharing" could encompass both.

**Comment by whit537** [comment]
As I'm implementing the per-tip overrides, it occurs to me that the implementation at https://github.com/gratipay/gratipay.com/pull/3112#issuecomment-72275370 has it that I'll be able to individually set who of my supporters is displayed publicly. I guess that happens on the receiving page? We can also email people when they get new supporters. I guess `public_supporters` encodes the default, then.

**Comment by whit537** [comment]
Or is that too complicated? Should we give both giver and receiver the ability to override individual tips? Or should we just give that ability to givers?

**Comment by whit537** [comment]
What if you require knowledge and then unrequire it? Do you get your former supporters back? Maybe within a certain timeframe?

**Comment by rohitpaulk** [comment]
> What if you require knowledge and then unrequire it? Do you get your former supporters back? Maybe within a certain timeframe?

I think the simplest way to implement this would be to handle it like we handle changes from a Team to Individual account. i.e When they click on 'require details', show them a warning that says that all their anonymous tips will be cancelled if they make this change.

**Comment by rummik** [comment]
@whit537 What I mean is that I feel like the abrupt question starting the section isn't very clear, and adds confusion to it, since we don't do that anywhere else

**Comment by synconme** [comment]
Just a suggestion regarding the options, maybe it is more convenient to simplify options a little bit, something like that:

Show other users:
* - [ ] My profile
* - [ ] Who I support
  * - [ ] Total amount I give
* - [ ] Who supports me
  * - [ ] Total amount I receive

-------
- [ ] Require my supporters tell me who they are and amount they give

From my point of view, if receiver wants to know who donates her, it is more likely she want to know the amount as well.

**Comment by whit537** [comment]
I'm scrapping this for parts.

**Pull request closed without merging**